swb.md

```diff
@@ -0,0 +1,46 @@
+The chakra implementation of concurrent GC and partial GC are based on background scanning the dirty pages. On windows, the OS kernel provides a feature called Write Watch( we also call it hardware write barrier), which the system keeps track of the pages that are written to in the committed memory region, ChakraCore GC use this information to determine which recycler allocated pages are dirty and do background rescan/mark on them. On non-Windows platform the hardware write barrier is not available, so we use software write barrier to track the dirty pages. In implementation of software write barrier ChakraCore uses a card table to map the whole address space, every page in the process has one byte in the card table corresponding to it. And all the recycler allocated non-Leaf struct have smart pointer wrapped to all its recycler pointer fields, while updating the pointer field, the smart pointer will update the corresponding byte in the card table, so the background thread knows the page is dirty. 
+
+The key to make the software write barrier work correctly and stably is to make sure all necessary pointer fields are annotated with the smart pointer correctly. In order to achieve this, here's a few rules:
+
+1. Annotate all fields recycler allocated struct/class with "Field" macro like below. Please also annotate non-pointer as well, it is a no-op for now but can use to provide more information for precised GC in the future. If you believe a pointer field will not be pointing to recycler memory, or, if it's pointing to recycler memory but the target recycler memory lifetime is tracked specifically on another struct/class, you can explicitly mark this field with FieldNoBarrier, while won't do barrier bit updating while modifying this field.
+```
+    struct A
+    {
+        Field(int)              intField;
+        Field(int*)             intArray;
+        Field(Var)              aVar;
+        Field(Field(Var)*)      varArray;
+        FieldNoBarrier(char*)   buffer; // e.g: buffer is heap allocated
+    }
+```
+
+2. If your function receive a recycler pointer to update, please use Field to define a local substitution of the argument and use that to do the update
+```
+    void UpdateSlot(Var* slots, int idx)
```

Comment:
Should we think more about this case... It is extremely easy to forget this. Should we ask people to try pass around "Field(Var)* slots" in the first place, instead of a cast? -- Line 21 needs a cast.